client/go/internal/vespa/deploy.go

```diff
@@ -119,6 +119,126 @@ func ZoneFromString(s string) (ZoneID, error) {
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return err
+	}
+	response, err := deployServiceDo(req, 0, deployment)
+	if err != nil {
+		return err
+	}
+	defer response.Body.Close()
+	_, err = io.Copy(w, response.Body)
+	return err
+}
+
+func fetchFromController(deployment DeploymentOptions, path string) error {
+	var (
+		pkgURL *url.URL
+		err    error
+	)
+	if deployment.Target.Deployment().Zone.Environment == "dev" {
+		pkgURL, err = deployment.url(fmt.Sprintf("/application/v4/tenant/%s/application/%s/instance/%s/job/%s/package",
+			deployment.Target.Deployment().Application.Tenant,
+			deployment.Target.Deployment().Application.Application,
+			deployment.Target.Deployment().Application.Instance,
+			deployment.Target.Deployment().Zone.Environment+"-"+deployment.Target.Deployment().Zone.Region,
+		))
+	} else {
+		pkgURL, err = deployment.url(fmt.Sprintf("/application/v4/tenant/%s/application/%s/package",
+			deployment.Target.Deployment().Application.Tenant,
+			deployment.Target.Deployment().Application.Application),
+		)
+	}
+	if err != nil {
+		return err
+	}
+	tmpFile, err := os.CreateTemp("", "vespa")
+	if err != nil {
+		return err
+	}
+	defer os.Remove(tmpFile.Name())
+	if err := deployServiceGet(pkgURL.String(), deployment, tmpFile); err != nil {
+		return err
+	}
+	if err := tmpFile.Close(); err != nil {
+		return err
+	}
+	return os.Rename(tmpFile.Name(), path)
+}
+
+func fetchFromConfigServer(deployment DeploymentOptions, path string) error {
+	tmpDir, err := os.MkdirTemp("", "vespa")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpDir)
+	u, err := deployment.url("/application/v2/tenant/default/application/default/environment/prod/region/default/instance/default/content")
+	if err != nil {
+		return err
+	}
+	dir := filepath.Join(tmpDir, "application")
+	if err := fetchFilesFromConfigServer(deployment, u, dir); err != nil {
+		return err
+	}
+	zipFile := filepath.Join(tmpDir, "application.zip")
+	if err := zipDir(dir, zipFile); err != nil {
+		return err
+	}
+	return os.Rename(zipFile, path)
+}
+
+func fetchFilesFromConfigServer(deployment DeploymentOptions, contentURL *url.URL, path string) error {
+	var data bytes.Buffer
+	if err := deployServiceGet(contentURL.String(), deployment, &data); err != nil {
+		return err
+	}
+	var fileURLs []string
+	if err := json.Unmarshal(data.Bytes(), &fileURLs); err != nil {
+		return err
+	}
+	for _, fu := range fileURLs {
+		u, err := url.Parse(fu)
+		if err != nil {
+			return err
+		}
+		entryName := filepath.Join(path, filepath.Base(u.Path))
+		if strings.HasSuffix(u.Path, "/") {
+			if err := fetchFilesFromConfigServer(deployment, u, entryName); err != nil {
+				return err
+			}
+		} else {
+			if err := os.MkdirAll(filepath.Dir(entryName), 0755); err != nil {
+				return err
+			}
+			f, err := os.Create(entryName)
+			if err != nil {
+				return err
+			}
+			if err := deployServiceGet(fu, deployment, f); err != nil {
```

Comment:
Any reason not to `defer f.close()`?

Reply:
This is called in a loop. If you defer closing, all files are kept open until the loop is done and the function returns. If the package has many files you could run into the open file limit.

Reply:
Ah, thanks for educating me :)